juju/juju
# [JUJU-4135] Rootless Juju controller on Kubernetes

This change sets up the juju controller statefulset to create juju controller pods that do not use **root** in any of the containers (including the charm container and init container).

To achieve this, we must emptyDir overlays via kubernetes to allow the various containers to:
- Offer introspection functions
- Offer juju tools such as juju-exec, juju-introspection and juju-dumplogs.
- Allow the juju processes to log to /var/log/juju

Here we declare the juju user and juju group have the uid/gid of `170`.

## QA steps

- bootstrap kubernetes controller
- check all the containers don't have root or mutable root filesystem
- check all processes for controller are running as 170 (juju user)
```
$ ps -u 170 -l
F S   UID     PID    PPID  C PRI  NI ADDR SZ WCHAN  TTY          TIME CMD
4 S   170 2598808 2598789  0  80   0 - 179278 -     ?        00:00:00 pebble
4 S   170 2598829 2598808  0  80   0 - 195865 -     ?        00:00:01 containeragent
4 S   170 2599125 2599105  0  80   0 - 179214 -     ?        00:00:00 pebble
4 S   170 2599278 2599258  1  80   0 - 459604 -     ?        00:00:03 mongod
0 S   170 2599597 2599125  0  80   0 - 228210 -     ?        00:00:01 jujud
```
- check introspection functions still work
```
juju ssh -m controller --container api-server 0 /bin/bash
$ whoami
juju
$ source /etc/profile.d/juju-introspection.sh
$ juju_controller_agent_name
controller-0
```
- deploying any charm should still have root

## Documentation changes

N/A

## Bug reference

N/A
